### PR TITLE
Packer, terraform and fabric files

### PR DESCRIPTION
**Pull request by alfraser86**
Here's the bits and bobs i've been working on.

There are lots of things to fix and refactor. But if the nginx config was tweaked for the node app then the packer and terraform would deploy the basic app and render it behind an elb within an autoscale group. (http://web-platform-elb-989836964.eu-west-1.elb.amazonaws.com/).

There is also a fabric file that we can use for continuing deploying the site with Jenkins. I have an idea of having the fabric file in the repo, checking it out to the jenkins server and then deploying with fabric, it should make the deployments really quick. What I haven't got quite right is the ssh keys to allow that, but i'll keep looking at that.

### CODE REVIEW

File: terraform/common-keys.tf [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+# new EC2 instances.
+#
+# Not so important since we manage SSH keys via the Packer AMI, but it is probably
+# still required for the provisioning process itself.
+#
+###############################
+
+resource "aws_key_pair" "ACF" {
+  key_name   = "ACF"
+  public_key = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAAAgQDE0i5YyRQAZmrVDe9p8h7/px4vg0yp26G82AaP9TjjgXOZ9xRv1glPYqgB0OAMmAu6G0czHAbFhxhOXQ5xZ+UEEGfVPImpgeL0+egjR1646MqKMjX/+5dSSf1NiLUTd4Ap5HyRqlCIHwID8NDU9ic0N50qz+rxv8At/Z1yr+6Bmw=="
```
Comment (jamesgorrie) [inline terraform/common-keys.tf]: As much as public keys aren't dangerous - it's best we keep them out. Using a more descriptive, portable name than "ACF" might be good to.

**Comment by jamesgorrie** [comment]
Just so that I understand the flow, we have an ELB that will then have instances deployed via terraform behind it? How are we pulling down old instances only once these are healthy?

This question is partly because it's hard to see what's what in the PR - any chance we could start to modularise the changes?

Having the Jenkins stuff in here is potentially problematic as I have been trying with Circle, which, from my perspective adds way less infrastructure, simplifying the deployments, and we don't have to worry about the very bespoke SSH server access.

File: terraform/common-keys.tf (hunk 1 above)
Comment (alfraser86) [inline terraform/common-keys.tf]: Yep, agreed. I'm not sure a public key is actually required for this to work. I added it purely for playing.

**Comment by alfraser86** [comment]
We can split it out. There are a few things.

The health check for the ELB defines when the old servers are pulled and the new ones come in. It checks port 80 to look for a healthy response and then switches them.

Only the Fab file is for jenkins, nothing else there is. 

I still favour using Jenkins for deployment over CircleCI as while there is a steeper learning curve it gives us ultimately more control over how and what we deploy. We also currently have a lot of sites all hopefully with continuous delivery and if we have all of them running both tests and deployments on CircleCI it would become fairly expensive. Saying that I could be convinced (https://www.slant.co/versus/2477/625/~jenkins_vs_circleci). 

 I agree that the SSH element is a bit of a pinch point.

**Comment by jamesgorrie** [comment]
Closing in favour of wellcome/web-platform-infra#1

**Pull request closed without merging**